File: ly_thuyet/buoi3(contest)/xau_ab_dai_n.cpp
#include<iostream>

using namespace std;

int main(){
    int t;
    cin>>t;
    while(t--){
        int n;
        cin>>n;
        string res="", last="";
        for(int i=0;i<n;i++){
            res+='A';
            last+='B';
        }
        cout<<res<<" ";
        while(res!=last){
            for(int i=res.length()-1;i>=0;i--){
                if(res[i]=='B'){
                    res[i]='A';
                    continue;
                }
                if(res[i]=='A'){
                    res[i]='B';
                    break;
                }
            }
            cout<<res<<' ';
        }
        cout<<endl;
    }
    return 0;
}

File: ly_thuyet/buoi1/xau_ab_dai_n.py
t = int(input())

for v in range(t):
    res, work, last = '', '', ''
    n = int(input())
    for i in range(n):
        work+='A'
        last+='B'
    res+=work+' '
    while work != last:
        l = n-1
        while l>=0:
            if(work[l]=='B'):
                strlst=list(work)
                strlst[l]='A'
                work="".join(strlst)
                l-=1
                continue
            if(work[l]=='A'):
                strlst=list(work)
                strlst[l]='B'
                work="".join(strlst)
                l-=1
                break
        res+=work+' '
    print(res)